lib/apis/v3/kubecontrollersconfig.go

```diff
@@ -0,0 +1,165 @@
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// KubeControllersConfiguration contains the configuration for Calico Kubernetes Controllers.
+type KubeControllersConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Specification of the KubeControllersConfiguration.
+	Spec KubeControllersConfigurationSpec `json:"spec,omitempty"`
+	// Status of the KubeControllersConfiguration
+	Status KubeControllersConfigurationStatus `json:"status,omitempty"`
+}
+
+// KubeControllersConfigurationSpec contains the values of the Kubernetes controllers configuration.
+type KubeControllersConfigurationSpec struct {
+
+	// LogSeverityScreen is the log severity above which logs are sent to the stdout. [Default: Info]
+	LogSeverityScreen string `json:"logSeverityScreen,omitempty" validate:"omitempty,logLevel"`
+
+	// HealthChecks enables or disables support for health checks
+	HealthChecks string `json:"healthChecks,omitempty" validate:"omitempty,oneof=Enabled Disabled"`
+
+	// EtcdV3CompactionPeriod is the period between etcdv3 compaction requests. Set to 0 to disable.
+	EtcdV3CompactionPeriod string `json:"etcdV3CompactionPeriod,omitempty" validate:"omitempty,duration"`
+
+	// Controllers enables and configures individual Kubernetes controllers
+	Controllers ControllersConfig `json:"controllers"`
+}
+
+// ControllersConfig enables and configures individual Kubernetes controllers
+type ControllersConfig struct {
+	// Node enables and configures the node controller
+	Node *NodeControllerConfig `json:"node,omitempty"`
+
+	// Policy enables and configures the policy controller
+	Policy *PolicyControllerConfig `json:"policy,omitempty"`
+
+	// WorkloadEndpoint enables and configures the workload endpoint controller
+	WorkloadEndpoint *WorkloadEndpointControllerConfig `json:"workloadEndpoint,omitempty"`
+
+	// ServiceAccount enables and configures the service account controller
+	ServiceAccount *ServiceAccountControllerConfig `json:"serviceAccount,omitempty"`
+
+	// Namespace enables and configures the namespace controller
+	Namespace *NamespaceControllerConfig `json:"namespace,omitempty"`
+}
+
+// NodeControllerConfig configures the node controller, which automatically cleans up configuration
+// for nodes that no longer exist. Optionally, it can create host endpoints for all Kubernetes nodes.
+type NodeControllerConfig struct {
+
+	// ReconcilerPeriod is the period to perform reconciliation with the Calico datastore.
+	ReconcilerPeriod string `json:"reconcilerPeriod,omitempty" validate:"omitempty,duration"`
+
+	// SyncLabels controls whether to copy Kubernetes node labels to Calico nodes
+	SyncLabels string `json:"syncLabels,omitempty" validate:"omitempty,oneof=Enabled Disabled"`
+
+	// HostEndpoint controls syncing nodes to host endpoints
+	HostEndpoint *AutoHostEndpointConfig `json:"hostEndpoint,omitempty"`
```

Comment:
WDYT about `HostEndpointConfig`?

Reply:
As the field name, I think it's redundant. It's embedded in a struct called "Config" so it's pretty clear, IMO.  Note that the _struct_ name includes "Config" since that symbol is out in the `v3` package namespace.

Reply:
I should have clarified, I meant the struct name. But I don't think it matters much.